pkg/utils/packet.go

```diff
@@ -196,3 +198,38 @@ func AnnounceIPs(ifName string, ipConfigs []*current.IPConfig) error {
 	}
 	return nil
 }
+
+// Blocking wait for interface ifName to have carrier (!NO_CARRIER flag).
+func WaitForCarrier(ifName string, waitTime time.Duration) bool {
+	var sleepDuration time.Duration
+
+	myNetLink := MyNetlink{}
+
+	start := time.Now()
+
+	for time.Since(start) < waitTime {
+		if sleepDuration == 0 {
+			sleepDuration = 2 * time.Millisecond
+		} else {
+			time.Sleep(sleepDuration)
+			/* Grow wait time exponentionally (factor 1.5). */
+			sleepDuration += sleepDuration / 2
+		}
+
+		linkObj, err := myNetLink.LinkByName(ifName)
+		if err != nil {
+			return false
+		}
+
+		/* Wait for carrier, i.e. IFF_UP|IFF_RUNNING. Note that there is also
+		 * IFF_LOWER_UP, but we follow iproute2 ([1]).
+		 *
+		 * [1] https://git.kernel.org/pub/scm/network/iproute2/iproute2.git/tree/ip/ipaddress.c?id=f9601b10c21145f76c3d46c163bac39515ed2061#n86
+		 */
+		if linkObj.Attrs().RawFlags&(unix.IFF_UP|unix.IFF_RUNNING) == (unix.IFF_UP | unix.IFF_RUNNING) {
```

Comment:
This will work if both IFF_UP and IFF_RUNNING flags are set. `IFF_RUNNING` seems to have a compatibility meaning only:

https://www.xml.com/ldd/chapter/book/ch14.html#t3
> IFF_RUNNING
This flag indicates that the interface is up and running. It is mostly present for BSD compatibility; the kernel makes little use of it. Most network drivers need not worry about IFF_RUNNING.

Is this the intended behavior?

Reply:
I think checking for IFF_UP|IFF_RUNNING is correct.

Maybe an alternative would be to check for IFF_LOWER_UP, but according to https://www.kernel.org/doc/html/v5.9/networking/operstates.html#querying-from-userspace , I think IFF_RUNNING (and IFF_UP) is more appropriate. That is also what iproute2 does (as linked).

Reply:
see for example: https://github.com/systemd/systemd/blob/77da8c4e9b167b1defe13896437389d8c219a926/src/shared/netif-util.c#L29

Reply:
Agree, thanks for adding details here